### PR TITLE
Add new leaderboards and matchmaker queues for 4v4

### PR DESCRIPTION
**Pull request by BlackYps**
I also set the 1v1 unit cap to 1500.
I didn't find anything for the map pools. Are they completely managed by the moderator client?

### CODE REVIEW

**Comment by Sheikah45** [comment]
Yes map pools are managed by the moderator client

**Comment by Sheikah45** [comment]
Also probably want to maybe start with the disabled so they don't show to people until the map pool is ready

**Comment by BlackYps** [comment]
I was thinking that we would deploy this when we want the queues to launch, but if we need the leaderboards first to set up the map pools, then yes, I can change the queues to be disabled.

**Comment by Sheikah45** [comment]
Yes you cannot set up the map pools until the queues exist

**Comment by Sheikah45** [comment]
Also here you did not add any matchmaker map pools so there will be nothing to add them to

**Comment by Sheikah45** [comment]
Actually looks like right now that is a manual brutus thing so probably need to add a flyaway for that as well

**Comment by BlackYps** [comment]
I added map pool migrations. I'm not too sure if this really is a good idea, because it could lead to less flexibility when trying to manage map pools from the moderator client. Maybe @Brutus5000 can shed some light on this.

**Comment by Sheikah45** [comment]
This only controls the rating range for the matchmaker brackets. Those haven't changed since we set the thing up.

The maps can still be set using the moderator client so it will not reduce flexibility from that perspective

**Comment by BlackYps** [comment]
I don't understand why the build fails with Unknown column 'matchmaker_queue_id'. Am I staring at some typo that I don't see?

**Comment by Sheikah45** [comment]
You are inserting into map_pool not matchmaker_queue_map_pool

File: migrations/R__090_matchmaker_queue_map_pool.sql [hunk 1]
```diff
@@ -0,0 +1,32 @@
+-- Repeatable migration: Flyway will replay this SQL every time this file (checksum) has changed!
+-- You can not delete items by removing them from this file! Also make sure never to edit or reuse existing ids!
+
+-- If you want to edit an existing or add a new matchmaker queue map pool, just append it to the SQL command and run flyway migrate.
+
+INSERT INTO `matchmaker_queue_map_pool` (id, matchmaker_queue_id, map_pool_id, min_rating, max_rating) VALUES
```
Comment (Brutus5000) [inline migrations/R__090_matchmaker_queue_map_pool.sql]: This table is managed via API. No flyway migrations needed/allowed.

File: migrations/R__080_map_pool.sql [hunk 2]
```diff
@@ -0,0 +1,29 @@
+-- Repeatable migration: Flyway will replay this SQL every time this file (checksum) has changed!
+-- You can not delete items by removing them from this file! Also make sure never to edit or reuse existing ids!
+
+-- If you want to edit an existing or add a new map pool, just append it to the SQL command and run flyway migrate.
+
+INSERT INTO `map_pool` (id, name) VALUES
```
Comment (Brutus5000) [inline migrations/R__080_map_pool.sql]: This table is managed via API. No flyway migrations needed/allowed.

File: migrations/R__060_leaderboard.sql [hunk 3]
```diff
@@ -7,7 +7,9 @@
 INSERT INTO `leaderboard` (id, initializer_id, technical_name, name_key, description_key) VALUES
 (1,NULL,'global','leaderboard.global.name','leaderboard.global.description'),
 (2,NULL,'ladder_1v1','leaderboard.ladder_1v1.name','leaderboard.ladder_1v1.description'),
-(3,1,'tmm_2v2','leaderboard.tmm_2v2.name','leaderboard.tmm_2v2.description')
+(3,1,'tmm_2v2','leaderboard.tmm_2v2.name','leaderboard.tmm_2v2.description'),
+(4,1,'tmm_4v4_full_share','leaderboard.tmm_4v4_full_share.name','leaderboard.tmm_4v4_full_share.description'),
+(5,1,'tmm_4v4_share_until_death','leaderboard.tmm_4v4_share_until_death.name','leaderboard.tmm_4v4_share_until_death.description')
```
Comment (Brutus5000) [inline migrations/R__060_leaderboard.sql:12]: Why do we need 2 different leaderboards? Skill is skill?

**Review by Brutus5000 — CHANGES_REQUESTED** [review]

File: migrations/R__060_leaderboard.sql (hunk 3 above)
Comment (BlackYps) [reply migrations/R__060_leaderboard.sql:12]: People argued that 4v4 full share and share until death are sufficiently different to warrant their own leaderboard. I'm not high rated enough to really have an informed opinion about this. I think Tagada can make the point better.

**Comment by Sheikah45** [comment]
@Brutus5000 From this https://github.com/FAForever/faf-moderator-client/issues/111#issuecomment-930515940

I took it that you wanted to just have them be managed by flyaway as well. Otherwise we may need to add the capability to the moderator client unless you just want people to do it through something like postman manually.

File: migrations/R__070_matchmaker_queue.sql [hunk 4]
```diff
@@ -4,9 +4,11 @@
...
 -- Matchmaker queues are handled dynamically. If they are enabled they will show up. (The map pool is managed by privileged users.)
 
-INSERT INTO `matchmaker_queue` (id, technical_name, featured_mod_id, leaderboard_id, name_key, team_size, enabled) VALUES
-(1,'ladder1v1',6,2,'matchmaker_queue.ladder1v1',1,1),
-(2,'tmm2v2',0,3,'matchmaker_queue.tmm2v2.name',2,1)
+INSERT INTO `matchmaker_queue` (id, technical_name, featured_mod_id, leaderboard_id, name_key, team_size, params, enabled) VALUES
+(1,'ladder1v1',6,2,'matchmaker_queue.ladder1v1',1,'UnitCap: 1500',1),
+(2,'tmm2v2',0,3,'matchmaker_queue.tmm2v2.name',2,NULL,1),
+(3,'tmm4v4FullShare',0,4,'matchmaker_queue.tmm4v4FullShare.name',4,NULL,0),
+(4,'tmm4v4ShareUntilDeath',0,5,'matchmaker_queue.tmm4v4ShareUntilDeath.name',4,'Share: ShareUntilDeath',0)
```
Comment (Askaholic) [inline migrations/R__070_matchmaker_queue.sql]: Imo we should stick to snake case

File: migrations/R__070_matchmaker_queue.sql [hunk 5]
```diff
@@ -4,9 +4,11 @@
 -- If you want to edit an existing or add a new matchmaker queue, just append it to the SQL command and run flyway migrate.
 -- Matchmaker queues are handled dynamically. If they are enabled they will show up. (The map pool is managed by privileged users.)
 
-INSERT INTO `matchmaker_queue` (id, technical_name, featured_mod_id, leaderboard_id, name_key, team_size, enabled) VALUES
-(1,'ladder1v1',6,2,'matchmaker_queue.ladder1v1',1,1),
-(2,'tmm2v2',0,3,'matchmaker_queue.tmm2v2.name',2,1)
+INSERT INTO `matchmaker_queue` (id, technical_name, featured_mod_id, leaderboard_id, name_key, team_size, params, enabled) VALUES
+(1,'ladder1v1',6,2,'matchmaker_queue.ladder1v1',1,'UnitCap: 1500',1),
```
Comment (Askaholic) [inline migrations/R__070_matchmaker_queue.sql]: Unrelated but when are we going to get rid of the ladder featured mod?

File: migrations/R__070_matchmaker_queue.sql (hunk 4 above)
Comment (BlackYps) [reply migrations/R__070_matchmaker_queue.sql]: I'll have to change it in the client as well, but I forgot the names for the new leaderboards anyway, so it's not that big of a deal.

File: migrations/R__070_matchmaker_queue.sql (hunk 5 above)
Comment (Sheikah45) [reply migrations/R__070_matchmaker_queue.sql]: I would vote just do it now

Comment (Sheikah45) [reply migrations/R__070_matchmaker_queue.sql]: I would vote just do it now

File: migrations/R__070_matchmaker_queue.sql [hunk 6]
```diff
@@ -4,9 +4,11 @@
...
 -- Matchmaker queues are handled dynamically. If they are enabled they will show up. (The map pool is managed by privileged users.)
 
-INSERT INTO `matchmaker_queue` (id, technical_name, featured_mod_id, leaderboard_id, name_key, team_size, enabled) VALUES
-(1,'ladder1v1',6,2,'matchmaker_queue.ladder1v1',1,1),
-(2,'tmm2v2',0,3,'matchmaker_queue.tmm2v2.name',2,1)
+INSERT INTO `matchmaker_queue` (id, technical_name, featured_mod_id, leaderboard_id, name_key, team_size, params, enabled) VALUES
+(1,'ladder1v1',0,2,'matchmaker_queue.ladder1v1',1,'UnitCap: 1500',1),
+(2,'tmm2v2',0,3,'matchmaker_queue.tmm2v2.name',2,NULL,1),
+(3,'tmm4v4_full_share',0,4,'matchmaker_queue.tmm4v4_full_share.name',4,NULL,0),
+(4,'tmm4v4_share_until_death',0,5,'matchmaker_queue.tmm4v4_share_until_death.name',4,'Share: ShareUntilDeath',0)
```
Comment (Askaholic) [inline migrations/R__070_matchmaker_queue.sql]: Params need to be valid JSON objects. See an example here:
https://github.com/FAForever/server/blob/c9c2e47ec47b0c9545b999256587b16267cded4b/tests/data/test-data.sql#L288

**Review by Askaholic — CHANGES_REQUESTED** [review]

**Comment by Brutus5000** [comment]
> @Brutus5000 From this [FAForever/faf-moderator-client#111 (comment)](https://github.com/FAForever/faf-moderator-client/issues/111#issuecomment-930515940)
> 
> I took it that you wanted to just have them be managed by flyaway as well. Otherwise we may need to add the capability to the moderator client unless you just want people to do it through something like postman manually.

I'm a difficult man, I don't know what I want.

**Review by Askaholic — APPROVED** [review]

**Comment by Askaholic** [comment]
Well, my original vision was that all of this would be set up through the mod client, but I suppose unless you’re the one implementing it you can’t expect to get what you want :P. And I never had the motivation to figure out how to develop on the mod client, although I think I did attempt to set it up once.

File: migrations/R__070_matchmaker_queue.sql [hunk 7]
```diff
@@ -4,9 +4,11 @@
 -- If you want to edit an existing or add a new matchmaker queue, just append it to the SQL command and run flyway migrate.
 -- Matchmaker queues are handled dynamically. If they are enabled they will show up. (The map pool is managed by privileged users.)
 
-INSERT INTO `matchmaker_queue` (id, technical_name, featured_mod_id, leaderboard_id, name_key, team_size, enabled) VALUES
-(1,'ladder1v1',6,2,'matchmaker_queue.ladder1v1',1,1),
-(2,'tmm2v2',0,3,'matchmaker_queue.tmm2v2.name',2,1)
+INSERT INTO `matchmaker_queue` (id, technical_name, featured_mod_id, leaderboard_id, name_key, team_size, params, enabled) VALUES
+(1,'ladder1v1',0,2,'matchmaker_queue.ladder1v1',1,'{"GameOptions":{"UnitCap":1500}}',1),
```
Comment (Askaholic) [inline migrations/R__070_matchmaker_queue.sql]: So I guess we don’t need the unit cap here after all… hooray!

https://github.com/FAForever/fa/pull/3533

**Comment by BlackYps** [comment]
Are we waiting for anything before merging this?

**Comment by Brutus5000** [comment]
No, I was just severely sick for 3 weeks.

**Pull request merged**